### PR TITLE
Coercion error path.

### PR DESCRIPTION
**Pull request by Warxcell**
I know that's not best place to put it, I'm looking for more elegant solution.

The idea is to have path of error in structured way, so we can format it to the frontend - and frontend can locate the error easier - otherwise frontend should parse strings to locate the error.

### CODE REVIEW

**Comment by spawnia** [comment]
I do not understand what this is trying to achieve. Please add a test that illustrates the effect of this change.

**Comment by Warxcell** [comment]
The idea is client of api to avoid parsing strings from error message like that
image

But directly receive path as array.

**Comment by spawnia** [comment]
We will implement nothing that breaks or extends the error format specified in the GraphQL specification, see https://spec.graphql.org/draft#sec-Errors. I think what you are trying to do will require an addition to `extensions`.

File: src/Utils/Value.php [hunk 1]
```diff
@@ -256,12 +259,22 @@ private static function coercionError(
...
+            } while ($path !== null);
+        }
+
         return new Error(
             $fullMessage,
             $blameNode,
             null,
             [],
-            null,
+            $atPath,
```
Comment (spawnia) [inline src/Utils/Value.php:277]: This is not what `$path` is intended for, it is meant to represent the path of fields leading up to the error - not arguments! https://github.com/webonyx/graphql-php/blob/6f4e702c810f2902e43671701fb5a3e5a48bb68f/src/Error/Error.php#L40-L46

**Comment by spawnia** [comment]
> Otherwise I should preg_match from error message

That might be the easiest. Alternatively, I could see wrapping the `$originalError` with an exception such as `CoercionError` and expose structured data there - or perhaps have `CoercionError` extend `Error`.

**Comment by Warxcell** [comment]
> 

Sounds perfect to me.

**Comment by spawnia** [comment]
Let's go with `class CoercionError extends Error` and reduce the number of nested exceptions.

File: src/Error/CoercionError.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
+<?php declare(strict_types=1);
+
+namespace GraphQL\Error;
+
+use GraphQL\Language\AST\Node;
+use GraphQL\Utils\Value;
+use Throwable;
+
+/**
+ * @phpstan-type Path list<string>
```
Comment (spawnia) [inline src/Error/CoercionError.php:10]: The path for values in an input list contains `int`. Also, let's differentiate this from the Path used for fields - how about `InputPath`? Same for the name of the property.

File: src/Utils/Value.php [hunk 3]
```diff
@@ -1,12 +1,10 @@
-<?php declare(strict_types=1);
+<?php
+
+declare(strict_types=1);
```
Comment (spawnia) [inline src/Utils/Value.php:3]: Please revert unnecessary formatting changes

File: src/Error/CoercionError.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        ?array $valuePath = null
+    ) {
+        parent::__construct($message, $nodes, null, [], null, $previous);
+        $this->valuePath = $valuePath;
+    }
+
+    /**
+     * @return Path|null
+     */
+    public function getValuePath(): ?array
```
Comment (spawnia) [inline src/Error/CoercionError.php:36]: Let's just make the property public

File: tests/Utils/CoerceValueTest.php [hunk 5]
```diff
@@ -76,17 +80,23 @@ public function testCoercingAnArrayToGraphQLIDProducesAnError(): void
      * Describe: for GraphQLInt.
      *
      * @param mixed $result returned result
+     * @phpstan-param Path|null $expectedPath
```
Comment (spawnia) [inline tests/Utils/CoerceValueTest.php:83]: Regular `@param` is fine here

File: tests/Utils/CoerceValueTest.php [hunk 6]
```diff
@@ -375,11 +385,11 @@ public function testReturnsErrorForIncorrectValueType(): void
     }
 
     /**
-     * @see it('returns no error for a valid input')
-     *
      * @param mixed $input
      *
      * @dataProvider validInputObjects
+     * @see          it('returns no error for a valid input')
```
Comment (spawnia) [inline tests/Utils/CoerceValueTest.php:391]: revert

**Review by spawnia — CHANGES_REQUESTED** [review]

File: src/Utils/Value.php (hunk 3 above)
Comment (Warxcell) [reply src/Utils/Value.php:3]: Formatting is PITA :( 
Every project should have commited their editorconfig / idea settings. Reverting

File: src/Error/CoercionError.php (hunk 2 above)
Comment (Warxcell) [reply src/Error/CoercionError.php:10]: ValuePath? Since its not only for input but also for enum && scalar.

Comment (spawnia) [reply src/Error/CoercionError.php:10]: Returned results are also values, but GraphQL distinguishes between input and output - see https://spec.graphql.org/draft/#sec-Input-and-Output-Types

Comment (Warxcell) [reply src/Error/CoercionError.php]: Renamed to InputPath

File: tests/Utils/CoerceValueTest.php [hunk 7]
```diff
@@ -76,17 +87,23 @@ public function testCoercingAnArrayToGraphQLIDProducesAnError(): void
...
         self::assertIsArray($errors);
         self::assertCount(1, $errors);
+        self::assertInstanceOf(CoercionError::class, $errors[0]);
         self::assertSame($expected, $errors[0]->getMessage());
 
         self::assertNull($result['value']);
+
+        if ($expectedPath !== null) {
+            self::assertSame($expectedPath, $errors[0]->inputPath);
+        }
```
Comment (spawnia) [inline tests/Utils/CoerceValueTest.php:106]: Shouldn't those be equivalent even when `null`?

File: tests/Utils/CoerceValueTest.php [hunk 8]
```diff
@@ -375,11 +392,11 @@ public function testReturnsErrorForIncorrectValueType(): void
     }
 
     /**
-     * @see it('returns no error for a valid input')
```
Comment (spawnia) [inline tests/Utils/CoerceValueTest.php:378]: please self-review and revert changes like this

File: tests/Utils/CoerceValueTest.php [hunk 9]
```diff
@@ -397,27 +414,32 @@ public function validInputObjects(): iterable
...
     /**
-     * @return iterable<int, array{mixed}>
+     * @return iterable<int, array{0: mixed, 1: string, 2: InputPath|null}>
      */
     public function invalidInputObjects(): iterable
     {
-        yield [123];
+        yield [['foo' => 1234, 'nested' => ['foobar' => null]], 'Expected non-nullable type Int! not to be null at value.nested.foobar.', ['nested', 'foobar']];
+        yield [['foo' => null, 'bar' => 1234], 'Expected non-nullable type Int! not to be null at value.foo.', ['foo']];
+        yield [123, 'Expected type TestInputObject to be an object.', null];
```
Comment (spawnia) [inline tests/Utils/CoerceValueTest.php:437]:
```suggestion
        yield [
            ['foo' => 1234, 'nested' => ['foobar' => null]],
            'Expected non-nullable type Int! not to be null at value.nested.foobar.',
            ['nested', 'foobar'],
        ];
        yield [
            ['foo' => null, 'bar' => 1234],
            'Expected non-nullable type Int! not to be null at value.foo.',
            ['foo'],
        ];
        yield [
            123,
            'Expected type TestInputObject to be an object.',
            null,
        ];
```

File: tests/Utils/CoerceValueTest.php [hunk 10]
```diff
@@ -397,27 +414,44 @@ public function validInputObjects(): iterable
...
     /**
-     * @return iterable<int, array{mixed}>
+     * @return iterable<int, array{0: mixed, 1: string, 2: InputPath|null}>
      */
     public function invalidInputObjects(): iterable
     {
-        yield [123];
+        yield [
+            ['foo' => 1234, 'nested' => ['foobar' => null]],
+            'Expected non-nullable type Int! not to be null at value.nested.foobar.',
```
Comment (spawnia) [inline tests/Utils/CoerceValueTest.php:437]: Where does `value` come from? Seems misleading to me.

File: tests/Utils/CoerceValueTest.php (hunk 8 above)
Comment (Warxcell) [reply tests/Utils/CoerceValueTest.php:378]: I can `vendor/bin/php-cs-fixer fix` before committing... seems its not working :shrug:

File: tests/Utils/CoerceValueTest.php (hunk 7 above)
Comment (Warxcell) [reply tests/Utils/CoerceValueTest.php:106]: Removed check.
